appengine/genetics/js/genetics.js

```diff
@@ -237,14 +513,507 @@ Genetics.init = function() {
+            new Genetics.Cage.Event('END_GAME', false).addToQueue();
+            return true;
+          }
+        } else if (event['TYPE'] != 'START_GAME' &&
+            event['TYPE'] != 'NEXT_ROUND') {
+          // If an unexpected event occurred, then the level is failed.
+          new Genetics.Cage.Event('END_GAME', false).addToQueue();
+          return true;
+        }
+      }
+      return false;
+    case 5:
+      // Player was asked to return a mouse that is a valid mate. The player
+      // succeeds if they mate 3 times successfully.
+      // fallthrough
+    case 6:
+      // Player was asked to return a mouse that is a valid mate and bigger
+      // than itself. The player succeeds if they mate 5 times successfully with
+      // a mate that fits this criteria.
+      var successfulMates = 0;
+      for (var i = 0, event; event = Genetics.Cage.Events[i]; i++) {
+        if (event['TYPE'] == 'MATE') {
+          if (event['RESULT'] == 'SUCCESS') {
+            if (BlocklyGames.LEVEL == 6 &&
+                Genetics.Cage.miceMap_[event['OPT_PARTNER']].size <=
+                Genetics.Mouse.SIZE) {
+              // If it is level 6 and the player tried to mate with someone
+              // that wasn't bigger than itself, they fail.
+              new Genetics.Cage.Event('END_GAME', false).addToQueue();
+              return true;
+            }
+            successfulMates++;
+            if (successfulMates == 5) {
+              new Genetics.Cage.Event('END_GAME', true).addToQueue();
+              return true;
+            }
+          } else {
+            // If the player had an unsuccessful mating attempt, then the level
+            // is failed.
+            new Genetics.Cage.Event('END_GAME', false).addToQueue();
+            return true;
+          }
+        } else if (event['TYPE'] != 'START_GAME' &&
+            event['TYPE'] != 'NEXT_ROUND') {
+          // If an unexpected event occurred, then the level is failed.
+          new Genetics.Cage.Event('END_GAME', false).addToQueue();
+          return true;
+        }
+      }
+      return false;
+    case 7:
+      // Player was asked to respond to mate request such that they return true
+      // if a mate is valid, false otherwise. The player succeeds if they
+      // correctly respond 5 times.
+      // fallthrough
+    case 8:
+      // Player was asked to respond to mate request such that they return true
+      // if a mate is valid and bigger than itself, false otherwise. The player
+      // succeeds if they correctly respond 5 times.
+      var successfulMates = 0;
+      for (var i = 0, event; event = Genetics.Cage.Events[i]; i++) {
+        if (event['TYPE'] == 'MATE') {
+          if (event['RESULT'] == 'SUCCESS') {
+            if (BlocklyGames.LEVEL == 8 &&
+                Genetics.Cage.miceMap_[event['ID']].size <=
+                Genetics.Mouse.SIZE) {
+              // If it is level 8 and the player tried to mate with someone
```

Comment:
I don't understand all these dependencies on "8", here and below. Aren't we in "case 8"?